go/internal/services/keys/validation.go

```diff
@@ -0,0 +1,243 @@
+package keys
+
+import (
+	"context"
+	"database/sql"
+	"fmt"
+	"strconv"
+	"strings"
+	"time"
+
+	"github.com/unkeyed/unkey/go/apps/api/openapi"
+	"github.com/unkeyed/unkey/go/internal/services/ratelimit"
+	"github.com/unkeyed/unkey/go/internal/services/usagelimiter"
+	"github.com/unkeyed/unkey/go/pkg/codes"
+	"github.com/unkeyed/unkey/go/pkg/fault"
+	"github.com/unkeyed/unkey/go/pkg/otel/tracing"
+	"github.com/unkeyed/unkey/go/pkg/prometheus/metrics"
+	"github.com/unkeyed/unkey/go/pkg/ptr"
+	"github.com/unkeyed/unkey/go/pkg/rbac"
+	"golang.org/x/exp/slices"
+)
+
+// withCredits validates that the key has sufficient usage credits and deducts the specified cost.
+// It updates the key's remaining request count and marks the key as invalid if the limit is exceeded.
+func (k *KeyVerifier) withCredits(ctx context.Context, cost int32) error {
+	ctx, span := tracing.Start(ctx, "verify.withCredits")
+	defer span.End()
+
+	if !k.Valid {
+		return nil
+	}
+
+	usage, err := k.usageLimiter.Limit(ctx, usagelimiter.UsageRequest{
+		KeyId: k.Key.ID,
+		Cost:  cost,
+	})
+	if err != nil {
+		return err
+	}
+
+	k.Key.RemainingRequests = sql.NullInt32{Int32: usage.Remaining, Valid: usage.Remaining != -1}
+	if !usage.Valid {
+		k.setInvalid(StatusUsageExceeded, "Key usage limit exceeded.")
+	}
+
+	// Emit Prometheus metrics for credits spent
+	identityID := ""
+	if k.Key.IdentityID.Valid {
+		identityID = k.Key.IdentityID.String
+	}
+	
+	// Credits are deducted when usage is valid AND cost > 0
+	deducted := usage.Valid && cost > 0
+	actualCostDeducted := int32(0)
+	if deducted {
+		actualCostDeducted = cost
+	}
+	
+	metrics.KeyCreditsSpentTotal.WithLabelValues(
+		k.AuthorizedWorkspaceID,           // workspace_id
+		k.Key.ID,                          // key_id
+		identityID,                        // identity_id
+		strconv.FormatBool(deducted),      // deducted - whether credits were actually deducted
+	).Add(float64(actualCostDeducted))     // Add the actual amount deducted, not the requested cost
+
+	return nil
+}
+
+// withIPWhitelist validates that the client IP address is in the key's IP whitelist.
+// If no whitelist is configured, this validation is skipped.
+func (k *KeyVerifier) withIPWhitelist() error {
+	if !k.Valid {
+		return nil
+	}
+
+	if !k.Key.IpWhitelist.Valid {
+		return nil
+	}
+
+	clientIP := k.session.Location()
+	if clientIP == "" {
+		k.Valid = false
+		k.Status = StatusForbidden
+		k.message = "client IP is required for IP whitelist validation"
+		return nil
+	}
+
+	allowedIPs := strings.Split(k.Key.IpWhitelist.String, ",")
```

Comment:
not a problem for now, but we should probably move this conversion to a place before it gets cached, so we don't do it on every verification

Reply:
@coderabbitai please create an issue for me